src/String.cc

```diff
@@ -163,8 +163,8 @@ String::append( char const *str, int len)
     assert(str && len >= 0);
 
     PROF_start(StringAppend);
-    if (len_ + len < size_) {
-        strncat(buf_, str, len);
+    if (len_ + len + 1 /*'\0'*/ < size_) {
+        xstrncpy(buf_+len_, str, len+1);
```

Comment:
I suggest adding a comment here to reduce the time developers will waste investigating this seemingly buggy (but working) code:

```
xstrncpy(buf_+len_, str, len+1); // xstrncpy() does not copy the last (n-th) byte
```

We should also adjust xstrncpy() documentation to emphasize its significant differences from standard strncpy(), but that adjustment can be considered outside this PR scope.

Reply:
You want to add such comments to every call of x*() code everywhere in Squid?
Or continue to expect dev to read the documentation of functions they use?

IMO it is more reasonable to expect the latter. Note the 'x' prefix _always_ indicates a locally modified version of the strncpy() API and/or behaviour - significant differences are expected for them. Functions we define for compat without API or behaviour differences use the standard name.

As you say documentation updates for the xstrncpy() are outside this PR scope. If you have a good idea what to write that it not already there please do PR or RFC that.